### DISCUSSION TITLE
Unhandled Rejection (Error):  auth0-spa-js must run on a secure origin.

### DISCUSSION THREAD

**Discussion by xiaoweizhaoEmmanuel** [l3atbc-datadog/pushkin#107] [General] [answered]
Hello,

I have been following the quick start and have encountered the unhandled rejection error during the "local testing" (though the site is on AWS). I run the pushkin site on an AWS EC2 instance, and after I typed in the IPv4 Public IP of my instance, the site tried to start, but failed with the error:

> Unhandled Rejection (Error): 
>       auth0-spa-js must run on a secure origin.
>       See https://github.com/auth0/auth0-spa-js/blob/master/FAQ.md#why-do-i-get-auth0-spa-js-must-run-on-a-secure-origin 
>       for more information.

    
I thought the problem might be the ip address was from http:// (which is not considered secure), so I added https:// in front of the IPV4 Public IP, but this time, the site refuse to connect.

Any ides? Many thanks!

**Comment by tmintz** [comment]
This happened to me when I entered the local IP address for my computer, but it worked when I used http://localhost instead.  Not sure why it would make a difference, but it worked for me.

**Reply by parkerrobbins** [reply]
@tmintz That's interesting. What operating system and browser(s) were you using?

**Reply by tmintz** [reply]
I'm using Chrome on macOS 10.13.6.

**Reply by parkerrobbins** [reply]
@tmintz Yes, I reproduced this on Ubuntu 18.04.4 with Chrome 84.0.4147.105. I cannot reproduce with Firefox 72.0.2, though, which works either way.

**Comment by parkerrobbins** [comment]
Hi, @xiaoweizhaoEmmanuel! Thanks for bringing this to our attention. 

Could you tell me what operating system and browser you're using? (If you have multiple browsers on your computer, do you have the same problem in all of them?)

**Reply by xiaoweizhaoEmmanuel** [reply]
I am using windows 10, and have the same problem for all the browsers.

**Reply by parkerrobbins** [reply]
@xiaoweizhaoEmmanuel Have you tried Firefox by any chance? Not a fix, more of a workaround for the moment while I try to reproduce your issue.

**Reply by xiaoweizhaoEmmanuel** [reply]
Many thanks! Yes, I have tried different browsers on different computers. Pushkin is run on an AWS instance (not my local machine), so I have to reach the Pushkin site by type in the ipv4 public ip, and it gives me the error message. 

Below is all the message I get:

> Unhandled Rejection (Error): 
      auth0-spa-js must run on a secure origin.
      See https://github.com/auth0/auth0-spa-js/blob/master/FAQ.md#why-do-i-get-auth0-spa-js-must-run-on-a-secure-origin 
      for more information.
    
(anonymous function)
src/src/utils.ts:354
  351 |     'For security reasons, `window.crypto` is required to run `auth0-spa-js`.'
  352 |   );
  353 | }
> 354 | if (typeof getCryptoSubtle() === 'undefined') {
      | ^  355 |   throw new Error(`
  356 |     auth0-spa-js must run on a secure origin.
  357 |     See https://github.com/auth0/auth0-spa-js/blob/master/FAQ.md#why-do-i-get-auth0-spa-js-must-run-on-a-secure-origin 
View compiled
new e
src/src/Auth0Client.ts:136
  133 | private worker: Worker;
  134 | 
  135 | constructor(private options: Auth0ClientOptions) {
> 136 |   typeof window !== 'undefined' && validateCrypto();
      | ^  137 |   this.cacheLocation = options.cacheLocation || CACHE_LOCATION_MEMORY;
  138 | 
  139 |   if (!cacheFactory(this.cacheLocation)) {
View compiled
(anonymous function)
src/src/index.ts:20
  17 | 
  18 | export default async function createAuth0Client(options: Auth0ClientOptions) {
  19 |   const auth0 = new Auth0Client(options);
> 20 |   await auth0.checkSession();
  21 |   return auth0;
  22 | }
  23 | 
View compiled
▶ 4 stack frames were collapsed.
(anonymous function)
src/src/index.ts:18
  15 | 
  16 | export * from './global';
  17 | 
> 18 | export default async function createAuth0Client(options: Auth0ClientOptions) {
  19 |   const auth0 = new Auth0Client(options);
  20 |   await auth0.checkSession();
  21 |   return auth0;
View compiled
initAuth0
src/app/src/utils/react-auth0-spa.js:34
  31 | 
  32 | useEffect(() => {
  33 |   const initAuth0 = async () => {
> 34 |     const auth0FromHook = await createAuth0Client(initOptions);
     | ^  35 |     setAuth0(auth0FromHook);
  36 | 
  37 |     if (
View compiled
(anonymous function)
src/app/src/utils/react-auth0-spa.js:56
  53 | 
  54 |     setLoading(false);
  55 |   };
> 56 |   initAuth0();
     | ^  57 |   // eslint-disable-next-line
  58 | }, []);
  59 | 
View compiled

**Reply by parkerrobbins** [reply]
@xiaoweizhaoEmmanuel I have reproduced your issue (although it works okay in Firefox for me). 

We are currently working on some changes to the login configuration, so stay tuned for those while we work on a more permanent solution that will also allow you to test logins.

@jkhartshorne, anything to add?

**Reply by jkhartshorne** [reply]
If you have authentication turned off and are using `https://...` rather than `http://...` it ought to work. But as @parkerrobbins says, we should have this a bit more streamlined shortly. Waiting for tests to complete on some related issues as I write this...

**Reply by MelindaPozzi** [reply]
I also have Windows 10 and I have the same issue for all the browsers.

I also get the same error message when I type my IPv4 Public IP address:

![Pushkin site](https://user-images.githubusercontent.com/33163405/90129776-8f145380-dd69-11ea-8a6a-352c4c21186f.JPG)
![Pushkin site2](https://user-images.githubusercontent.com/33163405/90129799-96d3f800-dd69-11ea-8f05-33e4d034d38f.JPG)

**Comment by jkhartshorne** [comment]
I'll be looking at this more this afternoon during the Q&A, time allowing. 

@parkerrobbins @restmegan -- can you please see if you can reproduce these issues?

**Reply by jkhartshorne** [reply]
@restmegan was able to reproduce this and is trying to figure it out.

**Reply by parkerrobbins** [reply]
@jkhartshorne 

I can reproduce this with my AWS instance as well. I am also not able to get around this on Firefox anymore, so I no longer have any methods to access my AWS instance's IPv4 Public IP.

You said above:
> If you have authentication turned off and are using https://... rather than http://... it ought to work. 

However, `https` is not possible when testing using an AWS instance (without installing an SSL certificate). When I turn authentication off and access my instance via `http`, I still get the error, but I'm not sure why that should happen, either. Are there any other places you know of that reference `auth0` other than `./pushkin/front-end/src/config.js`?

**Comment by jkhartshorne** [comment]
Can you please confirm that in `pushkin.yaml`, `useAuth` is set to `false`?

**Reply by xiaoweizhaoEmmanuel** [reply]
No. below is what is shown in my pushkin.yaml . I set auth0 to be false in ./pushkin/front-end/src/config.js , and then run pushkin prep again, still not working. Thanks!

```
# main directories relative to project root ('..')
experimentsDir: 'experiments'
coreDir: 'pushkin'

# databases configs experiments can use
databases:
  localtestdb:
    user: 'postgres'
    pass: 'example'
    url: 'localhost'
    name: 'test_db'

```

**Reply by parkerrobbins** [reply]
@jkhartshorne 

My `pushkin.yaml` looks the same. My understanding was that the `useAuth` setting lived in `./pushkin/front-end/src/config.js`, as @xiaoweizhaoEmmanuel said.

**Reply by gepping** [reply]
@parkerrobbins 
Thank you for your suggestion to use the IPv4 Public IP, my bad for not seeing that earlier.  I tried using both Google Chrome and Firefox to pull up the webpage but got the same Unhandled Rejection now.

**Reply by parkerrobbins** [reply]
@gepping I just re-added that info to the docs&mdash;totally my fault!

As far as this error goes, you are in the same boat as the rest of us at the moment. We are working on a solution. Thanks for your patience.

**Reply by gepping** [reply]
@parkerrobbins No worries and thank you for your help!

**Comment by jkhartshorne** [comment]
@gepping @parkerrobbins -- sorry, I was looking at the new, improved `pushkin.yaml`. You are right, in your version, it's in `config.js`. 

I can't reproduce your issues right now because I have the updated version of Pushkin running on my computer, and it's hard to run multiple versions simultaneously. I am hopefully very close to publishing that, along with a migration guide. (The changes you have to make are fairly minor.) Unfortunately, I don't have childcare right now, so things are slow. 

But I can confirm that in the current version, I don't seem to be having this problem. 

BTW is @restmegan having this issue?

**Reply by restmegan** [reply]
@jkhartshorne Yes I've also had this issue when using AWS instances, editing `config.js` doesn't resolve the issue. It seems like there are dependencies on auth0 in `pushkin/front-end/src/index.js` that could be causing this error

**Reply by parkerrobbins** [reply]
@jkhartshorne This issue can only be reproduced in one of two ways (that I know of):

1. Running an AWS instance and attempting to access the site through the instance's  IPv4 Public IP address.
2. Entering the computer's local IP address instead of `http://localhost` when testing locally

So if you aren't doing those things specifically, you won't reproduce the issue.

**Reply by jkhartshorne** [reply]
OK. So I think I understand what's going on here. Basically, auth0 requires you to have an https connection to the website UNLESS the website is running on the same computer you are using (ie, on 'localhost'). So that's why we're getting this refusal when running the site on an AWS server.

So one option is to buy an SSL certificate and learn how to install it. I don't think we want to do that. Longer-term, there may be a workaround. 

HOWEVER, what's causing this error is auth0, which is only used for authentication, which currently isn't supported by Pushkin anyway. So I'm going to release a new website template that has auth0 completely removed, rather than just deactivated. 

This means that in the future, Windows users would not be able to do local tests of websites using authentication. But we'll cross that bridge when we come to it. Also, my medium-term plans involve writing support for a staging website, which would mitigate this issue. (It's still nice to do local testing in addition to using a staging website, but if you can't, it's not the end of the world.)

I'll post here when the new template is ready.

**Comment by jkhartshorne** [comment, marked as answer]
@xiaoweizhaoEmmanuel @gepping @tmintz 

I think we have this fixed. This had to do with some security protections that are specific to doing development on an AWS EC2 container. 

The fix will require starting a new pushkin site from scratch. Either delete your old one -- or, if you have existing code that you don't want to abandon, start a new site in a new folder. I can help you figure out how to merge your old code into the new site. 

Then: 

1. Upgrade your CLI (if you use npm, that's `$ npm upgrade --global pushkin-cli --latest`)
2. Run `$ pushkin armageddon`
3. Redo the quickstart. But when choosing a site template, be sure to choose `pushkin-sitetemplates-basic`, NOT `pushkin-sitetemplates-default`.

**Reply by jkhartshorne** [reply]
Actually, we found one more bug. Working on it now.

**Reply by gepping** [reply]
@jkhartshorne 
Hi Josh, it's been a few weeks since your last comment so I wanted to check-in.  Were you able to fix that last bug?

**Reply by parkerrobbins** [reply]
Hi, @gepping! 

We have actually gotten Pushkin to work on Windows using the Windows Subsystem for Linux, which will allow you to do both local testing and AWS deploy. Using an EC2 instance is no longer the preferred method for testing/deploying Pushkin. Check out the new Windows instructions [here](https://languagelearninglab.gitbook.io/pushkin/getting-started/installing-pushkin-and-dependencies/windows-10/windows-install).

**Reply by gepping** [reply]
Hi @parkerrobbins, great to hear and thank you!

**Comment by restmegan** [comment]
Hi @xiaoweizhaoEmmanuel @tmintz @gepping ! I just wanted to check in to see how Pushkin has been going for you. The Pushkin team is online today answering questions so if there's anything you need help with please let us know!

**Reply by gepping** [reply]
Hi @restmegan !  I used the Windows Subsystem for Linux, as recommended by @parkerrobbins , and that worked perfectly.  I don't have any questions at this time, but thank you for checking in!

**Comment by tmintz** [comment]
Thanks for checking in.  Slammed with teaching, so no fun playing with Pushkin for weeks!

Best,
Toby
